### PR TITLE
Recipe of fp8 quant and autoquant for AM1.5

### PR DESCRIPTION
**Pull request by zewenli98**
This Recipe defines a reproducible PTQ procedure for quantizing Alpamayo 1.5 to FP8 or AutoQuant (FP8 + NVFP4). 
There are four new files including one 644-clip eval set, one 5k-clip calibration set, and two patches for Alpamayo 1.5 repo.

cc: @narendasan

### CODE REVIEW

**Comment by super-anova** [comment]
Hi. Could you please add pyprojec.toml and uv.lock for the recipe you added in this PR `recipes/alpamayo1_5_quant` (following the patterns of https://github.com/NVlabs/alpamayo-recipes/tree/main/recipes/alpamayo1_5_sft for example)? We have organized the recipes in a way that users can decide on which recipes to use for their use cases.

File: recipes/alpamayo1_5_quant/README.md [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+- `--calib_parquet <path>`: calibration clip source. `0417_5k_train_set_for_calibration_25.10.parquet` is used by default.
+- `--num_of_calib_clips <N>`: number of calibration clips (1 to 5000). `100` is used by default.
+
+Please refer to `eval.py` to find the usage of more arguments.
+
+### FP8 quantization
+
+Under the root of `alpamayo1.5` repo, run an example command below to quantize alpamayo1.5 in FP8 and save the quantized model:
+```bash
+PYTHONPATH=src python src/alpamayo1_5/eval.py --quant_format=fp8 --num_of_calib_clips=100 --save_model_dir=./models
```
Comment (super-anova) [inline recipes/alpamayo1_5_quant/README.md]: You are probably in the process of updating this readme but this file path is not correct

File: recipes/alpamayo1_5_quant/quantize.py [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+import argparse
+import time
+import pandas as pd
+from pathlib import Path
+from tqdm import tqdm
+import os
+
+import torch
+import numpy as np
+import modelopt.torch.opt as mto
```
Comment (super-anova) [inline recipes/alpamayo1_5_quant/quantize.py:24]: In the pyproject.toml for this recipe, make sure to list modelopt as a dependency.

**Comment by zewenli98** [comment]
Hi @super-anova Thanks for your review. I have added pyproject.toml and uv.lock. please take a look

File: recipes/alpamayo1_5_quant/README.md (hunk 1 above)
Comment (zewenli98) [reply recipes/alpamayo1_5_quant/README.md]: As https://github.com/NVlabs/alpamayo-recipes/pull/4/changes#diff-df9b1ab617a1e466d208c54a67af2ee98c6d63175bc0b242ca41528c07dd1d8cR54, the files in this PR are patches to [Alpamayo1.5](https://github.com/NVlabs/alpamayo1.5), so users need to go to Alpamayo1.5 to run eval.py.

**Comment by zewenli98** [comment]
Hi @super-anova Can you take a look if anything else need to be changed?

File: recipes/alpamayo1_5_quant/README.md [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+## Prerequisites
+This recipe is tested in the following settings. Other settings may also work but not guaranteed. 
+- NVIDIA B300 GPU with CUDA 13
+- Python 3.12
+- Python Libraries: torch==2.12.0, torchvision==0.27.0, nvidia-modelopt==0.43.0
+
+**NVIDIA Model Optimizer (ModelOpt)** is a library comprising state-of-the-art model optimization techniques including quantization and sparsity to compress models. It accepts a torch or ONNX model as input and provides Python APIs for users to easily stack different model optimization techniques to produce optimized & quantized checkpoints. In this recipe, we utilize ModelOpt to quantize Alpamayo 1.5.
+
+## Getting started
+### Download Alpamayo 1.5
```
Comment (super-anova) [inline recipes/alpamayo1_5_quant/README.md]: The entry point for this recipe should ideally be this recipe, instead of alpamayo 1.5 model repo.

More broadly, we recently added a contribution guideline document here: https://github.com/NVlabs/alpamayo-recipes/blob/main/CONTRIBUTING.md#recipe-contribution-guidelines

Could you please take a look and try to align with the high level guidance there? The design here is that a user will decide on which recipe is right for their use case, install that recipe, and do what they need using the recipe. More details are in the contributing.md

File: recipes/alpamayo1_5_quant/eval.py [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+
+    This avoids text-only calibration and ensures quantizers in the rollout path
+    (vlm/expert/diffusion-related modules) observe representative activations.
+    """
+    def _calibration_loop(runtime_model):
+        runtime_model.eval()
+        with torch.no_grad():
+            for clip_id in tqdm(clip_ids, desc="calibration"):
+                data = load_physical_aiavdataset(clip_id, t0_us=t0_us)
+                messages = helper.create_message(data["image_frames"].flatten(0, 1))
```
Comment (super-anova) [inline recipes/alpamayo1_5_quant/eval.py]: You also need `camera_indices` here. See example: https://github.com/NVlabs/alpamayo1.5/blob/main/src/alpamayo1_5/test_inference.py#L35

File: recipes/alpamayo1_5_quant/eval.py [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+    device: str = "cuda",
+    seed: int | None = 42,
+) -> tuple[float, float]:
+    """
+    Returns minADE (meters) for one clip.
+    """
+    data = load_physical_aiavdataset(clip_id, t0_us=t0_us)
+
+    # Build chat message and tokenize
+    messages = helper.create_message(data["image_frames"].flatten(0, 1))
```
Comment (super-anova) [inline recipes/alpamayo1_5_quant/eval.py]: Same comment here about camera indices

**Comment by narendasan** [comment]
@super-anova what are your thoughts on checking in the calibration and eval datasets, should these be pushed to HF somewhere or is it ok in repo?

**Comment by narendasan** [comment]
@super-anova Addressed the issues about `camera_indices`, using the recipe as the entry point and formatting it closer to the other recipes

File: recipes/alpamayo1_5_quant/utils.py [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+    Returns:
+        Quantized model (the search_state from mtq.auto_quantize is discarded).
+    """
+    from alpamayo1_5 import helper
+    from alpamayo1_5.load_physical_aiavdataset import load_physical_aiavdataset
+
+    def _one_epoch():
+        for clip_id in clip_ids:
+            data = load_physical_aiavdataset(clip_id, t0_us=t0_us)
+            messages = helper.create_message(data["image_frames"].flatten(0, 1))
```
Comment (super-anova) [inline recipes/alpamayo1_5_quant/utils.py]: do you need to pass in camera_indices here as well?

File: recipes/alpamayo1_5_quant/README.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+> **Tip:** If you hit HuggingFace Hub rate limits, set `export HF_HUB_OFFLINE=1`
+> and `export TRANSFORMERS_OFFLINE=1` to force all model/tokenizer loads from
+> local cache.
+
+| Variable                 | Required    | Purpose                                                                                  |
+| ------------------------ | ----------- | ---------------------------------------------------------------------------------------- |
+| `ALPAMAYO_WORKSPACE`     | yes         | Root of the `alpamayo-recipes` checkout                                                  |
+| `ALPAMAYO_MODEL_DIR`     | yes         | Pre-trained Alpamayo model directory (output of step 4)                                  |
+| `ALPAMAYO_PAI_LOCAL_DIR` | yes         | PAI dataset root (output of step 5); read by entry scripts at runtime                    |
+| `ALPAMAYO_LOG_DIR`       | yes         | Directory for Cosmos-RL logs                                                             |
```
Comment (super-anova) [inline recipes/alpamayo1_5_quant/README.md:85]: Is this ALPAMAYO_LOG_DIR env var required? It's not set in the README?

**Review by super-anova — COMMENTED** [review]
@narendasan Thanks for addressing my comments. I left two more comments.

Did you verify this recipe by following the README end to end from a fresh env? Does it only work for cuda 13 (I can't seem to get it to work on cluster nodes with cuda 12)?

**Comment by narendasan** [comment]
> @narendasan Thanks for addressing my comments. I left two more comments.
> 
> Did you verify this recipe by following the README end to end from a fresh env? Does it only work for cuda 13 (I can't seem to get it to work on cluster nodes with cuda 12)?

I verified end to end on my own system (DGX Spark). There is no specific requirement  I suspect the cuda 12 support is due to the pytorch version - I did not change this from @zewenli98's prior config, I can take a look deeper at this.

File: recipes/alpamayo1_5_quant/README.md (hunk 7 above)
Comment (narendasan) [reply recipes/alpamayo1_5_quant/README.md:85]: It might not be, my assumption was it was tied into `alpamayo_r1.common.logging`

**Comment by super-anova** [comment]
> > @narendasan Thanks for addressing my comments. I left two more comments.
> > Did you verify this recipe by following the README end to end from a fresh env? Does it only work for cuda 13 (I can't seem to get it to work on cluster nodes with cuda 12)?
> 
> I verified end to end on my own system (DGX Spark). There is no specific requirement I suspect the cuda 12 support is due to the pytorch version - I did not change this from @zewenli98's prior config, I can take a look deeper at this.

@narendasan Thanks. Let me know how that turns out. There are probably a decent number of developers out there who would benefit from supporting cuda 12.

**Comment by narendasan** [comment]
@super-anova Amended the pyproject.toml to be able to work both on cuda12 and cuda13. The default now is the same as the other recipes

**Comment by super-anova** [comment]
@narendasan Thanks! I verified that this now works with Cuda 12.

File: recipes/alpamayo1_5_quant/.gitignore [hunk 8]
```diff
@@ -0,0 +1,2 @@
+am15_quant/
```
Comment (super-anova) [inline recipes/alpamayo1_5_quant/.gitignore:1]: Is it better to move these lines to https://github.com/NVlabs/alpamayo-recipes/blob/main/.gitignore?

File: recipes/alpamayo1_5_quant/README.md [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+This Recipe defines a reproducible post-training quantization (PTQ) procedure for quantizing Alpamayo 1.5 to FP8 or AutoQuant (FP8 + NVFP4)
+
+## Prerequisites
+
+This recipe is tested in the following settings.
+
+- NVIDIA B300 GPU with CUDA 13
+- Python 3.12
+- Python Libraries: torch==2.8.0, torchvision==0.27.0, nvidia-modelopt==0.43.0
```
Comment (super-anova) [inline recipes/alpamayo1_5_quant/README.md]: in pyproject.toml, torchvision version is 0.23.0 instead?

File: recipes/alpamayo1_5_quant/README.md [hunk 10]
```diff
@@ -0,0 +1,177 @@
+# FP8 / AutoQuant (FP8 + NVFP4) Quantization
+
+This Recipe defines a reproducible post-training quantization (PTQ) procedure for quantizing Alpamayo 1.5 to FP8 or AutoQuant (FP8 + NVFP4)
+
+## Prerequisites
+
+This recipe is tested in the following settings.
+
+- NVIDIA B300 GPU with CUDA 13
```
Comment (super-anova) [inline recipes/alpamayo1_5_quant/README.md:10]: shall we add CUDA 12 and the corresponding hardware you verified to this README as well?

**Review by super-anova — COMMENTED** [review]
Thanks for the great work!

Probably my final questions: do we have any benchmark numbers on inference latency speedup and/or runtime memory reduction, etc., for the hardware configs that you tested? I imagine users will wonder if they are getting the expected benefits with quantization after using this recipe.

Also, I wonder if you have any tests available that we can add to this recipe? Our users may want to develop further on the recipe and would be very useful to have a test to test correctness etc.

File: recipes/alpamayo1_5_quant/README.md (hunk 10 above)
Comment (narendasan) [reply recipes/alpamayo1_5_quant/README.md:10]: Added the CU12 config

**Comment by narendasan** [comment]
> Probably my final questions: do we have any benchmark numbers on inference latency speedup and/or runtime memory reduction, etc., for the hardware configs that you tested? I imagine users will wonder if they are getting the expected benefits with quantization after using this recipe.

For latency I think we should probably run it through perflab if we are going to publicly release the numbers. We can start this effort internally and update the README later.

I have added the memory reduction numbers. 

> Also, I wonder if you have any tests available that we can add to this recipe? Our users may want to develop further on the recipe and would be very useful to have a test to test correctness etc.

What tests do you have in mind beyond verifying the accuracy difference between base and quantized? I would suspect the majority of changes that users would make would be in the model opt configuration for different quantization recipes. The rest of the code is mostly boilerplate

File: recipes/alpamayo1_5_quant/.gitignore (hunk 8 above)
Comment (narendasan) [reply recipes/alpamayo1_5_quant/.gitignore:1]: At least the venv is specific to this recipe. Up to you

File: recipes/alpamayo1_5_quant/README.md (hunk 9 above)
Comment (narendasan) [reply recipes/alpamayo1_5_quant/README.md]: ah yeah good catch

File: recipes/alpamayo1_5_quant/.gitignore (hunk 8 above)
Comment (super-anova) [reply recipes/alpamayo1_5_quant/.gitignore:1]: Cool let's leave it then.

File: recipes/alpamayo1_5_quant/utils.py (hunk 6 above)
Comment (super-anova) [reply recipes/alpamayo1_5_quant/utils.py]: Fixed now. Thanks!

File: recipes/alpamayo1_5_quant/quantize.py [hunk 11]
```diff
@@ -306,12 +313,22 @@ def main():
         )
     model.eval()
 
+    # Compress fake-quant (simulated QDQ) weights into the real low-bit format so the saved
```
Comment (super-anova) [inline recipes/alpamayo1_5_quant/quantize.py:321]: I tried to test eval using the compressed ckpt but I got errors
```
FAILED clip_id=...: self._dequantize is True and self.fake_quant is False. This case should have been handled.
No successful clips; average minADE not computed.
Failed clips: 1
```

Steps I followed:
```
uv run --active quantize.py \
  --quant_format=fp8 \
  --num_of_calib_clips=1 \
  --num_traj_samples=1 \
  --max_generation_length=16 \
  --save_model_dir="$YOUR_HOME/outputs"

uv run --active eval.py \
  --ckpt "$YOUR_HOME/outputs/alpamayo1.5_fp8_calib1" \
  --limit=1 \
  --num_traj_samples=1 \
  --max_generation_length=16 \
  --print_every=1
```

File: recipes/alpamayo1_5_quant/README.md [hunk 12]
```diff
@@ -6,11 +6,18 @@ This Recipe defines a reproducible post-training quantization (PTQ) procedure fo
...
 This recipe is tested in the following settings.
 
+- NVIDIA 5090 GPU with CUDA 12
 - NVIDIA B300 GPU with CUDA 13
 - Python 3.12
-- Python Libraries: torch==2.8.0, torchvision==0.27.0, nvidia-modelopt==0.43.0
+- Python Libraries: torch==2.8.0, torchvision==0.23.0, nvidia-modelopt==0.43.0
 
-**NVIDIA Model Optimizer (ModelOpt)** is a library comprising state-of-the-art model optimization techniques including quantization and sparsity to compress models. In this recipe, we utilize ModelOpt to quantize Alpamayo 1.5.
+**NVIDIA Model Optimizer (ModelOpt)** is a library comprising state-of-the-art model optimization techniques including quantization and sparsity to compress models. In this recipe, we utilize ModelOpt to post train quantize Alpamayo 1.5 losslessly.
```
Comment (super-anova) [inline recipes/alpamayo1_5_quant/README.md]: "losslessly" seems a pretty strong claim here without numbers backing it up. Maybe state something like "“with minimal accuracy loss”?

**Comment by super-anova** [comment]
> > Probably my final questions: do we have any benchmark numbers on inference latency speedup and/or runtime memory reduction, etc., for the hardware configs that you tested? I imagine users will wonder if they are getting the expected benefits with quantization after using this recipe.
> 
> For latency I think we should probably run it through perflab if we are going to publicly release the numbers. We can start this effort internally and update the README later.
> 
> I have added the memory reduction numbers.
> 
> > Also, I wonder if you have any tests available that we can add to this recipe? Our users may want to develop further on the recipe and would be very useful to have a test to test correctness etc.
> 
> What tests do you have in mind beyond verifying the accuracy difference between base and quantized? I would suspect the majority of changes that users would make would be in the model opt configuration for different quantization recipes. The rest of the code is mostly boilerplate

Makes sense to run inference latency numbers through perflab.

For tests, I was thinking something simple like asserting minADE is within some error tolerance on a sample clip. But I suppose not super super helpful as users can run the quantized model inference and see the minADE. We can skip this for now. Thanks.

**Review by super-anova — COMMENTED** [review]
Just a final comment on a failed attempt to run eval on the compressed ckpt. Thanks!

File: recipes/alpamayo1_5_quant/quantize.py (hunk 11 above)
Comment (narendasan) [reply recipes/alpamayo1_5_quant/quantize.py:321]: There seems to be some issue with conv and the compressed ckpt. Left it out for very little penalty (should still work in fake quant mode for TensorRT). Will follow up with ModelOpt but should pass now.

File: recipes/alpamayo1_5_quant/README.md (hunk 12 above)
Comment (narendasan) [reply recipes/alpamayo1_5_quant/README.md]: changed

**Comment by super-anova** [comment]
@narendasan Thanks for addressing all the final comments! All looks good to me now.
@BorisIvanovic We can merge the PR if it also looks good to you.

**Review by super-anova — APPROVED** [review]

**Review by BorisIvanovic — APPROVED** [review]
Looks great, excited about this one!
